Added template for Alcatel AOS: `show interfaces port`

evilmonkey19 opened the pull request.

mjbear commented on tests/alcatel_aos/show_interfaces_port/alcatel_aos_show_interfaces_port_R6.raw:
@evilmonkey19 
Might you have or be able to get raw data where the `#` Wait To Restore is in effect?

This would improve the test data! Thank you.

evilmonkey19 replied:
Unfortunately I don't :(. Either I know how to make it work. It is an option that i know it exists but I don't use it at all. If it happens that you know hot to make it work I have equipment to get the output :)

mjbear replied:
> Unfortunately I don't :(. Either I know how to make it work. It is an option that i know it exists but I don't use it at all. If it happens that you know hot to make it work I have equipment to get the output :)

Page 1-33 (or search for `interfaces wait-to-restore`)
```
interfaces {slot chassis/slot| port chassis/slot/port[-port2]} wait-to-restore num
```
https://www.al-enterprise.com/-/media/assets/internet/documents/omniswitch-aos-release-810r1-cli-reference-user-guide-rev-a-en.pdf

@evilmonkey19
From the sounds of it, once you set a "wait to restore" timer on a (spare?) port, cause the port to shutdown (I would think the easiest is to administratively shut the port down from the CLI then grab output.

Let me know how that goes! :grinning:

mjbear commented on ntc_templates/templates/alcatel_aos_show_interfaces_port.textfsm:
```diff
@@ -0,0 +1,16 @@
+Value VIOLATIONS (\S+)
+Value RECOVERY_TIME (\d+)
+Value RECOVERY_MAX (\d+)
+Value WTR (\d+)
+Value ALIAS (\S+)
+
+Start
+  ^\s*[-+]+\s*$$
+  ^\s*${WTR_RUNNING}${PERMANENT_SHUTDOWN}${PORT}\s+${ADMIN_STATUS}\s+${LINK_STATUS}\s+${VIOLATIONS}\s+${RECOVERY_TIME}\s+${RECOVERY_MAX}\s+${WTR}\s+${ALIAS}\s*$$ -> Record
```
```suggestion
Value WTR_TIME (\d+)
Value ALIAS (\S+)

Start
  ^\s*-+\+
  ^\s*${WTR_RUNNING}${PERMANENT_SHUTDOWN}${PORT}\s+${ADMIN_STATUS}\s+${LINK_STATUS}\s+${VIOLATIONS}\s+${RECOVERY_TIME}\s+${RECOVERY_MAX}\s+${WTR_TIME}\s+${ALIAS}\s*$$ -> Record
```

Use the separator pattern that @jmcgill298 supports.

Improve self-description of capture group by making `WTR` >> `WTR_TIME`.

Thank you!

mjbear commented on tests/alcatel_aos/show_interfaces_port/alcatel_aos_show_interfaces_port_R6.yml:
```diff
@@ -0,0 +1,72 @@
+  - admin_status: "disable"
+    alias: "\"\""
+    link_status: "down"
+    permanent_shutdown: ""
+    port: "1/23"
+    recovery_max: "10"
+    recovery_time: "300"
+    violations: "none"
+    wtr: "0"
+    wtr_running: ""
+  - admin_status: "enable"
+    alias: "\"\""
+    link_status: "up"
+    permanent_shutdown: ""
+    port: "1/25"
+    recovery_max: "10"
+    recovery_time: "300"
+    violations: "none"
+    wtr: "0"
+    wtr_running: ""
+  - admin_status: "enable"
+    alias: "\"\""
+    link_status: "down"
+    permanent_shutdown: ""
+    port: "1/26"
+    recovery_max: "10"
+    recovery_time: "300"
+    violations: "none"
+    wtr: "0"
+    wtr_running: ""
+  - admin_status: "enable"
+    alias: "\"\""
+    link_status: "down"
+    permanent_shutdown: ""
+    port: "1/28"
+    recovery_max: "0"
+    recovery_time: "0"
+    violations: "none"
+    wtr: "0"
+    wtr_running: ""
```
Updated parsed data for the `WTR` >> `WTR_TIME` change.

evilmonkey19 commented:
I commited your sugggestion @mjbear and I also fixed a small bug with the alias (the ""). Do you have any suggestion on how to catch it in a more elegant way?

mjbear commented:
> I commited your sugggestion @mjbear and I also fixed a small bug with the alias (the ""). Do you have any suggestion on how to catch it in a more elegant way?

Hello!
By the looks of the raw output, the pair of double quotes are there regardless if there's alias text or not.
I'd guess you could make one of those other lines `"Hello World"`. **Two words with spaces. This is possible, right?**

I'll figure yes and at that point we can make the ALIAS regex capture spaces, but also make it optional. There are at least two ways, but I'll stick with the simplest by using `.*` for ALIAS. Look out for my edit where I anchor with the double quotes that seem to always be there. :grin:

On **WTR**, I'll have to look at the manual and we try to get it to happen. Do you have AOS devices that in dev/test environment? :stuck_out_tongue_winking_eye:

jmcgill298 commented on ntc_templates/templates/alcatel_aos_show_interfaces_port.textfsm:
```diff
@@ -0,0 +1,16 @@
+Start
+  ^\s*-+\+
+  ^\s*${WTR_RUNNING}${PERMANENT_SHUTDOWN}${PORT}\s+${ADMIN_STATUS}\s+${LINK_STATUS}\s+${VIOLATIONS}\s+${RECOVERY_TIME}\s+${RECOVERY_MAX}\s+${WTR_TIME}\s+"${ALIAS}"\s*$$ -> Record
+  ^\s*
```
```suggestion
  ^\s*$$
```
We don't want to just catch any line that stars with a space, since that would negate our purpose with having the Error line. We had some cases before we implemented the Error line policy where templates where silently ignoring significant lines because something in the data was different than we had seen before, and our regex didn't account for that. In order to prevent that from happening, we want to have something significant to match on for each line, or else raise an error.

evilmonkey19 commented:
Implementing the `\s*` has the issue that the tests don't work. How can i make them work in this case?

evilmonkey19 commented:
> @evilmonkey19 Another thing - could you try to get Wait To Restore output?
> 
> 🔥 Here's my notes from earlier. [#1908 (comment)](https://github.com/networktocode/ntc-templates/pull/1908#discussion_r1856674751)

This is a raw output using the commands you suggest :)
```
Legends: WTR - Wait To Restore
         #   - WTR Timer is Running & Port is in wait-to-restore state
         *   - Permanent Shutdown

Slot/    Admin     Link    Violations  Recovery   Recovery      WTR            Alias
Port     Status   Status                 Time       Max        (sec)
------+----------+---------+----------+----------+----------+----------+-----------------------------------------
  1/1    enable      up        none           300         10          0 ""
  1/2    enable      down      none           300         10          0 ""
  1/3    enable      down      none           300         10          0 ""
  1/4    enable      down      none           300         10          0 ""
  1/5    enable      down      none           300         10          0 ""
  1/6    enable      down      none           300         10          0 ""
  1/7    enable      down      none           300         10          0 ""
  1/8    enable      down      none           300         10          0 ""
  1/9    enable      down      none           300         10          0 ""
  1/10   enable      down      none           300         10          0 ""
  1/11   enable      down      none           300         10       # 10 ""
  1/12   enable      down      none           300         10          0 ""
  1/13   enable      down      none           300         10          0 ""
  1/14   enable      down      none           300         10          0 ""
  1/15   enable      down      none           300         10          0 ""
  1/16   enable      down      none           300         10          0 ""
  1/17   enable      down      none           300         10          0 ""
  1/18   enable      down      none           300         10          0 ""
  1/19   enable      down      none           300         10          0 ""
  1/20   enable      down      none           300         10          0 ""
  1/21   enable      down      none           300         10          0 ""
  1/22   enable      down      none           300         10          0 ""
  1/23   enable      down      none           300         10          0 ""
  1/24   enable      down      none           300         10          0 ""
  1/25   enable      down      none           300         10          0 ""
  1/26   enable      down      none           300         10          0 ""
  1/27   enable      down      none             0          0          0 ""
  1/28   enable      down      none             0          0          0 ""
```

jmcgill298 commented on ntc_templates/templates/alcatel_aos_show_interfaces_port.textfsm:
```diff
@@ -0,0 +1,18 @@
+Value ALIAS (.*)
+
+Start
+  ^\s*-+\+ -> Table
```
```suggestion
  ^\s*-+\+ -> Table
  ^Legends: WTR - Wait To Restore
  ^\s+#
  ^\s+*
  ^\s*Slot/\s+Admin\s+Link\s+Violations\s+Recovery\s+Recovery\s+WTR\s+Alias\s*$$
  ^\s*Port\s+Status\s+Status\s+Time\s+Max\s+(sec)\s*$$
```

I like to match the table headers to ensure we are getting the headers we expect

evilmonkey19 replied:
Looks good to me!

mjbear commented on ntc_templates/templates/alcatel_aos_show_interfaces_port.textfsm:
```diff
@@ -1,16 +1,18 @@
-  ^\s*-+\+
-  ^\s*${WTR_RUNNING}${PERMANENT_SHUTDOWN}${PORT}\s+${ADMIN_STATUS}\s+${LINK_STATUS}\s+${VIOLATIONS}\s+${RECOVERY_TIME}\s+${RECOVERY_MAX}\s+${WTR_TIME}\s+"${ALIAS}"\s*$$ -> Record
+  ^\s*-+\+ -> Table
+
```
@jmcgill298
Should the Error rule/directive be added to the `Start` state for good measure?

```suggestion
  ^.*$$ -> Error

```

mjbear commented:
@evilmonkey19
I thought we changed `PORT` to be simpler `\S+` similar to other templates. I'm going to put that suggestion forth.

There are fairly minor merge conflicts (I ran through fixing them just to see).

* reordering lines in the `index` since more AOS templates were added
* a couple of merge conflict resolutions in the textfsm template itself

Not horrible, but an extra obstacle before the automatic tests for this project run.
Since I rebased on a feature branch to test the conflict resolution this means commit hashes change. Whether I put forth a PR to you or you fix the conflicts in your branch by rebasing ... it will likely mean at least one force push from you.

mjbear commented on ntc_templates/templates/alcatel_aos_show_interfaces_port.textfsm:
```diff
@@ -0,0 +1,25 @@
+Value PORT ((\d+\/?)+)
```
Yep, I'm backtracking to a simpler regex. This also matches the PORT regex used by other AOS templates.

```suggestion
Value PORT (\S+)
```

evilmonkey19 commented:
> @evilmonkey19 I thought we changed `PORT` to be simpler `\S+` similar to other templates. I'm going to put that suggestion forth.
> 
> There are fairly minor merge conflicts (I ran through fixing them just to see).
> 
>     * reordering lines in the `index` since more AOS templates were added
> 
>     * a couple of merge conflict resolutions in the textfsm template itself
> 
> 
> Not horrible, but an extra obstacle before the automatic tests for this project run. Since I rebased on a feature branch to test the conflict resolution this means commit hashes change. Whether I put forth a PR to you or you fix the conflicts in your branch by rebasing ... it will likely mean at least one force push from you.

No problem :) I can do the force push for sure

jmcgill298 commented on ntc_templates/templates/alcatel_aos_show_interfaces_port.textfsm:
```diff
@@ -0,0 +1,25 @@
+Value PORT ((\d+\/?)+)
+Value PERMANENT_SHUTDOWN (\*?)
```
I don't think it matters, it should have the same result
